polyval/src/field/backend/pclmulqdq.rs

```diff
@@ -0,0 +1,108 @@
+//! Support for the VPCLMULQDQ CPU intrinsic on `x86` and `x86_64` target
+//! architectures.
+
+#[cfg(target_arch = "x86")]
+use core::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use core::arch::x86_64::*;
+
+use super::Xmm;
+use crate::field::clmul::{self, Clmul};
+use core::ops::{BitXor, BitXorAssign};
+
+/// 2 x `u64` values loaded into a `__m128i` register
+#[repr(align(16))]
+#[derive(Copy, Clone)]
+pub struct U64x2(__m128i);
+
+impl From<[u64; 2]> for U64x2 {
+    fn from(array: [u64; 2]) -> U64x2 {
+        unsafe { _mm_loadu_si128(array.as_ptr() as *const __m128i) }.into()
+    }
+}
+
+impl From<U64x2> for [u64; 2] {
+    fn from(u64x2: U64x2) -> [u64; 2] {
+        let mut result = [0u64; 2];
+
+        unsafe {
+            _mm_storeu_si128(result.as_mut_ptr() as *mut __m128i, u64x2.0);
+        }
+
+        result
+    }
+}
+
+impl From<__m128i> for U64x2 {
+    fn from(mm: __m128i) -> U64x2 {
+        U64x2(mm)
+    }
+}
+
+impl From<U64x2> for __m128i {
+    fn from(u64x2: U64x2) -> __m128i {
+        u64x2.0
+    }
+}
+
+impl BitXor for U64x2 {
+    type Output = Self;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        U64x2(unsafe { xor(self.0, rhs.0) })
+    }
+}
+
+impl BitXorAssign for U64x2 {
+    fn bitxor_assign(&mut self, rhs: Self) {
+        // TODO(tarcieri): optimize
+        self.0 = unsafe { xor(self.0, rhs.0) };
+    }
+}
+
+impl Clmul for U64x2 {
+    fn clmul<I>(self, rhs: Self, imm: I) -> Self
+    where
+        I: Into<clmul::PseudoOp>,
+    {
+        unsafe { vpclmulqdq(self.0, rhs.0, imm.into()) }.into()
+    }
+}
+
+// TODO(tarcieri): optimized `rotate_left`, `shift_right`, `shift_left`
+impl Xmm for U64x2 {}
+
+#[target_feature(enable = "sse2", enable = "sse4.1")]
+unsafe fn xor(a: __m128i, b: __m128i) -> __m128i {
+    _mm_xor_si128(a, b)
+}
+
+#[target_feature(enable = "pclmulqdq", enable = "sse2", enable = "sse4.1")]
+unsafe fn vpclmulqdq(a: __m128i, b: __m128i, op: clmul::PseudoOp) -> __m128i {
```

Comment:
@newpavlov somewhat successfully using the stuff you added way back here 😉 https://github.com/rust-lang/stdarch/pull/320/files